aptos-move/move-examples/rewards_pool/sources/rewards_pool.move

```diff
@@ -0,0 +1,288 @@
+module rewards_pool::rewards_pool {
+    use aptos_framework::fungible_asset::{Self, FungibleAsset, FungibleStore, Metadata};
+    use aptos_framework::primary_fungible_store;
+    use aptos_framework::object::{Self, Object, ExtendRef};
+    use aptos_std::pool_u64_unbound::{Self as pool_u64, Pool};
+    use aptos_std::simple_map::{Self, SimpleMap};
+    use aptos_std::smart_table::{Self, SmartTable};
+
+    use rewards_pool::epoch;
+
+    use std::signer;
+    use std::vector;
+
+    /// Rewards can only be claimed for epochs that have ended.
+    const EREWARDS_CANNOT_BE_CLAIMED_FOR_CURRENT_EPOCH: u64 = 1;
+    /// The rewards pool does not support the given reward token type.
+    const EREWARD_TOKEN_NOT_SUPPORTED: u64 = 2;
+    /// The caller doesn't have any pending rewards.
+    const ENO_REWARDS_TO_CLAIM: u64 = 3;
+
+    /// Data regarding the rewards to be distributed for a specific epoch.
+    struct EpochRewards has store {
+        /// Total amount of rewards for each reward token added to this epoch.
+        total_amounts: SimpleMap<Object<Metadata>, u64>,
+        /// Pool representing the claimer shares in this epoch.
+        claimer_pool: Pool,
+    }
+
+    /// Data regarding the store object for a specific reward token.
+    struct RewardStore has store {
+        /// The fungible store for this reward token.
+        store: Object<FungibleStore>,
+        /// We need to keep the fungible store's extend ref to be able to transfer rewards from it during claiming.
+        store_extend_ref: ExtendRef,
+    }
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    struct RewardsPool has key {
+        /// A mapping to track per epoch rewards data.
+        epoch_rewards: SmartTable<u64, EpochRewards>,
+        /// The stores where rewards are kept.
+        reward_stores: SimpleMap<Object<Metadata>, RewardStore>,
+    }
+
+    /// Create a new rewards pool with the given reward tokens (fungible assets only)
+    public entry fun create_entry(reward_tokens: vector<Object<Metadata>>) {
+        create(reward_tokens);
+    }
+
+    /// Create a new rewards pool with the given reward tokens (fungible assets only)
+    public fun create(reward_tokens: vector<Object<Metadata>>): Object<RewardsPool> {
+        // The owner of the object doesn't matter as there are no owner-based permissions.
+        // If developers want to be extra cautious here, they can make the owner @0x0.
+        // Here the reward pool also doesn't keep an ExtendRef so there would be no way to obtain its signer.
+        let rewards_pool_constructor_ref = &object::create_object(@rewards_pool);
+        let rewards_pool_signer = &object::generate_signer(rewards_pool_constructor_ref);
+        let rewards_pool_addr = signer::address_of(rewards_pool_signer);
+        let reward_stores = simple_map::new();
+        vector::for_each(reward_tokens, |reward_token| {
+            let reward_token: Object<Metadata> = reward_token;
```

Comment:
is this a deficiency in inline functions?

Reply:
Yep. It doesn't infer types correctly sometimes. There's a bug tracking this https://github.com/aptos-labs/aptos-core/issues/6922 but @wrwg mentioned this would not be fixed in compiler v1